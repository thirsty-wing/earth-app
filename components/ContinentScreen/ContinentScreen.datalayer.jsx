import useFetchEarth from "../../utils/hooks/useFetchEarth";
import ContinentScreen from "./ContinentScreen";

/**
 * Shows the continent screen with data
 *
 * @param {object} props
 * @param {object} props.route
 * @param {object} props.route.params
 * @param {string} props.route.params.code
 * @param {object} props.navigation
 * @returns {JSX.Element}
 */
function ContinentScreenDatalayer(props) {

  const {
    route,
    navigation,
  } = props;

  const code = route?.params?.code;

  const [
    data,
    responseStatus,
  ] = useFetchEarth(
    `query{
      continent(code: "${code}") {
        name
        code
        countries {
          name
          code
          emoji
        }
      }
    }`
  );

  return(
    <ContinentScreen
      data={data}
      responseStatus={responseStatus}
      navigation={navigation}
    />
  );
}

export default ContinentScreenDatalayer;
