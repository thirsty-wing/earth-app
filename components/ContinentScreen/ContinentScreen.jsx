import { ActivityIndicator, ScrollView, Text, View, Pressable } from "react-native";

import BackgroundView from "../BackgroundView";
import styles from "./ContinentScreen.styles";


/**
 * Shows the continent screen
 *
 * @param {object} props
 * @param {object} [props.data]
 * @param {"LOADING"|"ERROR"|"OK"} [props.responseStatus = "OK"]
 * @param {object} [props.router]
 * @returns {JSX.Element}
 */
function ContinentScreen(props) {

  const {
    data,
    responseStatus = "OK",
    router,
  } = props;


  return(
    <BackgroundView>
      <ScrollView style={styles.scrollView}>

        { responseStatus === "LOADING" &&
          <ActivityIndicator size="large" />
        }


        { responseStatus === "ERROR" &&
            <Text style={styles.plainText}>
              Something went wrong
            </Text>
        }


        { responseStatus === "OK" &&
          <>
            <View style={styles.topView}>
              <Pressable
                style={styles.backPressable}
                onPress={router?.back}
              >
                <Text style={styles.plainText}>
                  Back
                </Text>
              </Pressable>
              <View style={styles.titleView}>
                <Text style={styles.titleText}>
                  {data?.continent?.name}
                </Text>
              </View>
            </View>
            <Text style={styles.plainText}>
              Two letter code: {data?.continent?.code}
            </Text>
            <Text style={styles.plainText}>
              Number of countries: {data?.continent?.countries?.length}
            </Text>
          </>
        }

      </ScrollView>
    </BackgroundView>
  );
}

export default ContinentScreen;
