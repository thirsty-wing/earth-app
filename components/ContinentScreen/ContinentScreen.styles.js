import { StyleSheet } from "react-native";

import { theme } from "../../utils/theme";


const styles = StyleSheet.create({
  countryView: {
    width: "100%",
    backgroundColor: theme.palette.action.primary,
    padding: theme.spacing(),
    display: "flex",
    flexDirection: "row",
    alignItems: "center",
    gap: theme.spacing(),
    borderRadius: 16,
  },
  scrollView: {
    gap: theme.spacing(),
    padding: theme.spacing(3),
    flex: 1,
  },
  showDataView: {
    gap: theme.spacing(),
    flex: 1,
  },
  plainText: {
    color: theme.palette.text.primary,
  },
  topView: {
    display: "flex",
    width: "100%",
    justifyContent: "flex-start",
    flexDirection: "row",
    paddingHorizontal: theme.spacing(3),
  },
  backPressable: {
    padding: theme.spacing(2),
    backgroundColor: theme.palette.action.primary,
    borderRadius: 16,
  },
  emojiText: {
    fontSize: theme.sizes.text.title,
  },
  subsectionText: {
    color: theme.palette.text.primary,
    fontSize: theme.sizes.text.subsection,
    flexWrap: "wrap",
    flex: 1,
  },
  svgView: {
    backgroundColor: theme.palette.text.primary,
    padding: theme.spacing(),
    display: "flex",
    justifyContent: "center",
    alignItems: "center",
    position: "absolute",
    top: 0,
    bottom: 0,
    left: 0,
    right: 0,
    borderRadius: 16,
  },
  svgViewWidthWrapper: {
    width: "100%",
  },
  svgViewHeightWrapper: {
    paddingBottom: "100%",
  },
  titleView: {
    flex: 1,
    justifyContent: "center",
    flexDirection: "row",
  },
  titleText: {
    color: theme.palette.text.primary,
    fontSize: theme.sizes.text.section,
  },
});

export default styles;
