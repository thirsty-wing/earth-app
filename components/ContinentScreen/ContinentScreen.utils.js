/**
 * Given an array of countries, return the number of unique languages
 *
 * @param {object[]} [countries]
 * @returns {number}
 */
export function numberOfLanguages(countries) {

  return new Set( // remove duplicates
    countries?.map( // in each country
      country => country?.languages?.map( // in each language
        language => language?.code // get the language code
      )
    )
      .flat() // flatten the array of arrays containing language codes
      .filter(languageCode => !!languageCode) // remove falsy values
  ).size; // return number of languages
}

