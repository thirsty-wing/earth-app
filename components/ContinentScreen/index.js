import ContinentScreen from "./ContinentScreen.datalayer";

export default ContinentScreen;
