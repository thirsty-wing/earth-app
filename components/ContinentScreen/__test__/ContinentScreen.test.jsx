import ContinentScreen from "../ContinentScreen";
import { shallow } from "../../../utils/test/shallow";


jest.mock("react-native-safe-area-context", () => ({
  useSafeAreaInsets: () => ({
    left: 0,
    right: 0,
    top: 0,
    bottom: 0,
  }),
}));


it("should render", () => {
  const content = shallow(<ContinentScreen/>);

  expect(content).toMatchSnapshot();
});

it("should render with LOADING response status", () => {
  const content = shallow(<ContinentScreen responseStatus="LOADING"/>);

  expect(content).toMatchSnapshot();
});

it("should render with ERROR response status", () => {
  const content = shallow(<ContinentScreen responseStatus="ERROR"/>);

  expect(content).toMatchSnapshot();
});
