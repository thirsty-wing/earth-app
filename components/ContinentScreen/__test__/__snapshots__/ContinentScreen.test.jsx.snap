// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
<BackgroundView>
  <View
    style={
      {
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "flex-start",
        "paddingHorizontal": 24,
        "width": "100%",
      }
    }
  >
    <Memo(Pressable)
      style={
        {
          "backgroundColor": "#049",
          "padding": 16,
        }
      }
    >
      <Text
        style={
          {
            "color": "#EEF",
          }
        }
      >
        Back
      </Text>
    </Memo(Pressable)>
    <View
      style={
        {
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          {
            "color": "#EEF",
            "fontSize": 30,
          }
        }
      />
    </View>
  </View>
  <ScrollView
    style={
      {
        "flex": 1,
        "gap": 8,
        "padding": 24,
      }
    }
  >
    <React.Fragment>
      <View
        style={
          {
            "width": "100%",
          }
        }
      >
        <View
          style={
            {
              "paddingBottom": "100%",
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#EEF",
                "borderRadius": 16,
                "bottom": 0,
                "display": "flex",
                "justifyContent": "center",
                "left": 0,
                "padding": 8,
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
          >
            <ContinentSvg />
          </View>
        </View>
      </View>
      <Text
        style={
          {
            "color": "#EEF",
            "fontSize": 25,
          }
        }
      >
        Stats
      </Text>
      <Text
        style={
          {
            "color": "#EEF",
          }
        }
      >
        Two letter code: 
      </Text>
      <Text
        style={
          {
            "color": "#EEF",
          }
        }
      >
        Number of countries: 
      </Text>
      <Text
        style={
          {
            "color": "#EEF",
          }
        }
      >
        Number of languages: 
        0
      </Text>
      <Text
        style={
          {
            "color": "#EEF",
            "fontSize": 25,
          }
        }
      >
        Countries
      </Text>
    </React.Fragment>
  </ScrollView>
</BackgroundView>
`;

exports[`should render with ERROR response status 1`] = `
<BackgroundView>
  <View
    style={
      {
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "flex-start",
        "paddingHorizontal": 24,
        "width": "100%",
      }
    }
  >
    <Memo(Pressable)
      style={
        {
          "backgroundColor": "#049",
          "padding": 16,
        }
      }
    >
      <Text
        style={
          {
            "color": "#EEF",
          }
        }
      >
        Back
      </Text>
    </Memo(Pressable)>
    <View
      style={
        {
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          {
            "color": "#EEF",
            "fontSize": 30,
          }
        }
      />
    </View>
  </View>
  <ScrollView
    style={
      {
        "flex": 1,
        "gap": 8,
        "padding": 24,
      }
    }
  >
    <Text
      style={
        {
          "color": "#EEF",
        }
      }
    >
      Something went wrong
    </Text>
  </ScrollView>
</BackgroundView>
`;

exports[`should render with LOADING response status 1`] = `
<BackgroundView>
  <View
    style={
      {
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "flex-start",
        "paddingHorizontal": 24,
        "width": "100%",
      }
    }
  >
    <Memo(Pressable)
      style={
        {
          "backgroundColor": "#049",
          "padding": 16,
        }
      }
    >
      <Text
        style={
          {
            "color": "#EEF",
          }
        }
      >
        Back
      </Text>
    </Memo(Pressable)>
    <View
      style={
        {
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          {
            "color": "#EEF",
            "fontSize": 30,
          }
        }
      />
    </View>
  </View>
  <ScrollView
    style={
      {
        "flex": 1,
        "gap": 8,
        "padding": 24,
      }
    }
  >
    <ActivityIndicator
      size="large"
    />
  </ScrollView>
</BackgroundView>
`;
